api/client/request.go

```diff
@@ -61,10 +62,71 @@ func (r *Request) HTTPRequest() (*http.Request, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Set GetBody so posted data is preserved on redirects or requests with
+	// expired access tokens.
+	req.GetBody = r.GetBody
 	req.Header = r.Header
 	return req, nil
 }
 
+type recordReader struct {
+	io.Reader
+	buf      *bytes.Buffer
+	limit    int
+	noreplay bool
+}
+
+func (r *recordReader) Read(b []byte) (int, error) {
+	n, err := r.Reader.Read(b)
+	if remaining := r.limit - r.buf.Len(); remaining > 0 {
+		cc := n
+		if n > remaining {
+			cc = remaining
+		}
+		r.buf.Write(b[:cc])
+	} else {
+		// Set noreplay to true so we know that replay has exceeded the buffer
+		// range and therefore the request cannot be replayed.
+		r.noreplay = true
+	}
+	return n, err
+}
+
+type replayReader struct {
+	io.Reader
+	buf *bytes.Reader
+}
+
+func (r *replayReader) Read(b []byte) (int, error) {
+	if r.buf.Len() > 0 {
+		return r.buf.Read(b)
+	}
+	return r.Reader.Read(b)
+}
+
+func (r *Request) GetBody() (io.ReadCloser, error) {
```

Comment:
Nits: I don't think this is really part of the public interface, so I wouldn't export it. I also think it's more interesting than `recordReader` and `replayReader` and deserves to come first.